source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,339 @@
+Snapshot reads are supported both on primaries and secondaries.
+
+MongoDatabase changes
+=====================
+
+There are no additional API changes to ``MongoDatabase`` beyond those specified in
+the Sessions Specification. All ``MongoDatabase`` methods that talk to the server
+have been overloaded to take a session parameter. If that session was started
+with ``isSnapshot = true`` then all read operations using that session will
+will share the same snapshot.
+
+MongoCollection changes
+=======================
+
+There are no additional API changes to ``MongoCollection`` beyond those specified
+in the Sessions Specification. All ``MongoCollection`` methods that talk to the
+server have been overloaded to take a session parameter. If that session was
+started with ``isSnapshot = true`` then all operations using that
+session will share the same snapshot.
+
+ReadConcern changes
+===================
+
+``snapshot`` added to `ReadConcernLevel enumeration <../read-write-concern/read-write-concern.rst#read-concern>`_.`.
+
+Server Commands
+===============
+
+There are no new server commands related to snapshot reads. Instead,
+snapshot reads are implemented by:
+
+1. Saving the ``atClusterTime`` returned by 5.0+ servers for the first find/aggregate operation in a
+   property ``snapshotTimestamp`` of the ``ClientSession`` object. Drivers MUST save the ``atClusterTime``
+   in the ``ClientSession`` object.
+
+2. Passing that ``snapshotTimestamp`` in the ``atClusterTime`` field of the ``readConcern`` field
+   for subsequent snapshot read operations (for find/aggregate/distinct commands).
+
+Server Command Responses
+========================
+
+To support snapshot reads the server returns the ``atClusterTime`` in
+cursor object it sends to the driver (for both find/aggregate commands).
+
+.. code:: typescript
+
+    {
+        ok : 1 or 0,
+        ... // the rest of the command reply
+        cursor : {
+            ... // the rest of the cursor reply
+            atClusterTime : <BsonTimestamp>
+        }
+    }
+
+For distinct commands server returns the ``atClusterTime`` in
+distinct response object it sends to the driver.
+
+.. code:: typescript
+
+    {
+        ok : 1 or 0,
+        ... // the rest of the command reply
+        atClusterTime : <BsonTimestamp>
+    }
+
+The ``atClusterTime`` MUST be stored in the ``ClientSession`` to later be passed as the
+``atClusterTime`` field of the ``readConcern`` with ``snapshot`` level field  in subsequent read operations.
+
+Server Errors
+=============
+1. The server may reply to read commands with a ``SnapshotTooOld`` error if the client's ``atClusterTime`` value is not available in the server's history.
+2. The server will return ``InvalidOptions`` error if both ``atClusterTime`` and ``afterClusterTime`` options are set to true.
+
+Snapshot read commands
+======================
+
+For snapshot reads the driver MUST first obtain ``atClusterTime`` from cursor response of find/aggregate command,
```

Comment:
Same question about `distinct` here. Can only `find` or `aggregate` fetch the first `atClusterTime` to use in subsequent read operations?